### PR TITLE
Enable dynamic replacement with library evolution

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://problem/58457716>.

### CODE REVIEW

File: lib/IRGen/Linking.cpp [hunk 1]
```diff
@@ -629,6 +629,8 @@ SILLinkage LinkEntity::getLinkage(ForDefinition_t forDefinition) const {
     return SILLinkage::Private;
 
   case Kind::DynamicallyReplaceableFunctionKey:
+    return getSILFunction()->getLinkage();
```
Comment (slavapestov) [inline lib/IRGen/Linking.cpp:632]: @aschwaighofer This was the only actual change I had to make. Are you aware of anything that I might have missed?

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
The way we choose which accessors get marked dynamically replaceable is not right. JohnM and Jrose explained to me that we should use the opaque accessors instead of whatever I choose to replace.

**Comment by slavapestov** [comment]
@aschwaighofer Do you have an example that we miscompile today?

@rjmccall Do you remember what the issue was?

**Comment by aschwaighofer** [comment]
I don't have an example. But the accessors that are printed in the swiftinterface file need to be the ones that are marked dynamic, I think.

**Comment by aschwaighofer** [comment]
Here is the comment from the commit 1980c4af3cd101a8205969af5a998698da601f59 that disable dynamic in library evolution mode:

```
Disallow `dynamic` on native Swift declarations under library-evolution mode
    
    The current ABI of dynamic is not what we want for certain declarations.
    
    e.g
    
    dynamic var x : Int {
      willset {
      }
    }
    
    Is exposed in the .swiftinterface file as
    
    dynamic var x : Int {
      get {}
      set {}
    }
    
    However, the actual implementation is marking the willSet only dynamic.
    
    var x : Int {
      dynamic willSet() {}
      get {}
      set {}
    }
    
    To be able to rely on `dynamic` in the future we disable dynamic (for
    non-@objc) under library-evolution for now until we have made sure that
    the ABI is proper.
    
    rdar://51678074
```

**Comment by aschwaighofer** [comment]
I am not sure that we actually print all opaque accessors in the swift interface file.
I feel like _modify accessor is also part of the set of opaque accessors?

**Comment by slavapestov** [comment]
@aschwaighofer I don't understand _why_, though. There's no resilience boundary between the instrumented library and the dynamic replacement stub.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
I'm going to merge this PR since it fixes the class method linkage issue. I'll keep poking at the accessor thing to try to come up with an example where the current scheme fails.

**Comment by aschwaighofer** [comment]
> I don't understand why, though. There's no resilience boundary between the instrumented library and the dynamic replacement stub.

This might be the case how it is being used today. But dynamic is a language feature like any other. So a resilient dynamic foo should be resilient. It should be possible to resiliently replace it.
I can't think of a reason why it can't be so.
You might be limited in what you do in the replacement but it should be possible?

If we ship the compiler with 'invalid' interface files we can rely on them without some sort of versioning in the future (or so was the argument brought up in the radar).

I am not 100% sure anymore but the example above I think also is a distillation of a problem I ran into at some point. The mismatch was causing a problem somehow: I can see this causing problems when the compilation of the replacement would use the .swiftinterface file. It would try to replace the wrong entry points.

**Comment by aschwaighofer** [comment]
It might be worth running the resilience tests with -enable-implicit-dynamic enabled and to see that does not cause issues.

I ran the all tests like this at one point (hack the compiler to have enable-implicit-dynamic enabled is probably the easiest way to do this).

**Comment by aschwaighofer** [comment]
> The mismatch was causing a problem somehow: I can see this causing problems when the compilation of the replacement would use the .swiftinterface file. It would try to replace the wrong entry points.

This might no longer reproduce today because this could have only reproduced when we were (erroneously) preferring .swiftinterface over .swiftmodule files?

**Comment by compnerd** [comment]
@slavapestov - this seems to have caused a regression: https://ci-external.swift.org/job/swift-PR-windows/1596/consoleText

**Pull request merged**